### ISSUE TITLE
Catch-all variant constructor for a tag_field

### ISSUE THREAD

**Issue by cyberhuman** [mjambon/atdgen#31]
Hi,
I would like  to do something like
```
type variant = [ A of a | B of b | _ of json ]
value <json tag_field="type"> : variant
```
Basically, to have a catch-all constructor for unknown tag field types.
Does atdgen support this?

**Comment by dsheets** [comment]
I believe the current support is limited to default tag field values with `json` type. That is:

```
type tag = [ A | B | Json ]
type variant = [ A of a | B of b | Json of json ]
type default_record = {
  ~tag <ocaml default="`Json"> : tag;
  variant <json tag_field="tag"> : variant;
}
```

With these types, any JSON will be accepted under `variant` field but only if `tag` is missing or `"Json"`. If `tag` is `"A"` or `"B"`, `variant` field will be restricted appropriately.

It's also possible to define related sets of variants, some of which are `json`-valued by reusing the `tag_field` with a different value field type (`[A of json | B of json | Json of json]`) so that no matter what the tag field value, the value field value will carry the variant tag and an arbitrary JSON payload.

I had originally planned to write a general-purpose fallback like you describe when I made my PR last autumn (see <https://github.com/mjambon/atdgen/pull/25#issuecomment-62309955>) but there were a number of complications and I didn't *need* support so I dropped it. I believe `yojson` 1.2.0+ is now fully capable to support the feature but maybe 10-15hrs of work is needed on `atdgen` to add support. It would certainly be useful for binding to only a fragment of a JSON API. :-/
